Coding Style (#97)
javascript-tutorial/cs.javascript.info: 1-js/03-code-quality/02-coding-style/article.md

> # Styl kódování

Reviewer: ```suggestion
# Styl zápisu programu
```

Náš kód musí být co nejčistší a co nejsnadněji čitelný.

> To je v podstatě umění programování -- vzít složitou úlohu a zakódovat ji tak, aby kód byl současně správný a čitelný člověkem. Dobrý styl kódování tomu značně napomáhá.

Reviewer: ```suggestion
To je v podstatě umění programování -- vzít složitou úlohu a naprogramovat ji tak, aby kód byl současně správný a čitelný člověkem. Dobrý styl zápisu tomu značně napomáhá.
```

Například místo použití vnořené podmínky `if` tímto způsobem:

```js
for (let i = 0; i < 10; i++) {
  if (podmínka) {
    ... // <- o jednu úroveň vnoření víc
  }
}
```

můžeme napsat:

```js
for (let i = 0; i < 10; i++) {
  if (!podmínka) *!*continue*/!*;
  ...  // <- žádná další úroveň vnoření
}
```

Podobně můžeme postupovat s `if/else` a `return`.

Například dvě níže uvedené konstrukce jsou identické.

Možnost 1:

```js
function mocnina(x, n) {
  if (n < 0) {
    alert("Záporné 'n' není podporováno");
  } else {
    let výsledek = 1;

    for (let i = 0; i < n; i++) {
      výsledek *= x;
    }

    return výsledek;
  }  
}
```

Možnost 2:

```js
function mocnina(x, n) {
  if (n < 0) {
    alert("Záporné 'n' není podporováno");
    return;
  }

  let výsledek = 1;

  for (let i = 0; i < n; i++) {
    výsledek *= x;
  }

  return výsledek;
}
```

Druhá konstrukce je čitelnější, protože „zvláštní případ“ `n < 0` je ošetřen hned na začátku. Když je kontrola provedena, můžeme se přesunout k „hlavnímu“ toku kódu, aniž bychom potřebovali další vnoření.

## Umístění funkcí

> Jestliže píšeme několik „pomocných“ funkcí a kód, který je používá, máme tři možnosti, jak funkce organizovat.

Reviewer: ```suggestion
Jestliže píšeme několik „pomocných“ funkcí a kód, který je používá, tak máme tři možnosti, jak funkce organizovat.
```

## Styloví průvodci

Stylový průvodce obsahuje obecná pravidla o tom, „jak psát“ kód, tj. které uvozovky používat, o kolik mezer odsazovat, maximální délku řádku atd. Mnoho vedlejších věcí.

Když všichni členové týmu používají stejného stylového průvodce, bude kód vypadat jednotně bez ohledu na to, který člen týmu ho napsal.

Samozřejmě si tým vždy může napsat svého vlastního stylového průvodce, ale obvykle to není zapotřebí. Existuje mnoho průvodců, z nichž si lze vybrat.

Některé oblíbené možnosti:

- [JavaScriptový stylový průvodce Google](https://google.github.io/styleguide/jsguide.html)
- [JavaScriptový stylový průvodce Airbnb](https://github.com/airbnb/javascript)
- [Idiomatic.JS](https://github.com/rwaldron/idiomatic.js)
- [StandardJS](https://standardjs.com/)
- (a mnoho dalších)

Pokud jste začínající vývojář, začněte podle příkladu na začátku této kapitoly. Pak si můžete projít jiné stylové průvodce, abyste získali další nápady a rozhodli se, který se vám líbí nejvíc.

## Automatické lintery

Lintery jsou nástroje, které umějí automaticky zkontrolovat styl vašeho kódu a navrhnout zlepšení.

> Výborné na nich je, že tato kontrola stylu umí najít i některé chyby, například překlepy v názvech proměnných nebo funkcí. Pro tuto jejich vlastnost se používání linteru doporučuje, i když se nechcete vázat k jednomu konkrétnímu „kódovacímu stylu“.

Reviewer: ```suggestion
Výborné na nich je, že tato kontrola stylu umí najít i některé chyby, například překlepy v názvech proměnných nebo funkcí. Pro tuto jejich vlastnost se používání linteru doporučuje, i když se nechcete vázat k jednomu konkrétnímu „stylu zápisu“.
```

}
```

Zde direktiva `"extends"` oznamuje, že konfigurace je založena na sadě nastavení „eslint:recommended“. Poté si specifikujeme vlastní.

Je také možné si místo toho stáhnout sady stylových pravidel z webu a pak je rozšířit. Pro podrobnosti ohledně instalace viz <https://eslint.org/docs/user-guide/getting-started>.

Také některá IDE obsahují zabudovaný linting, který je vyhovující, ale ne tak nastavitelný jako ESLint.

## Shrnutí

Všechna syntaktická pravidla uvedená v této kapitole (a v odkazovaných stylových průvodcích) směřují ke zvýšení čitelnosti vašeho kódu. Všechna jsou diskutovatelná.

Když uvažujeme o psaní „lepšího“ kódu, měli bychom si klást otázky „Co učiní kód čitelnějším a snadnějším k porozumění?“ a „Co nám pomůže vyvarovat se chyb?“. To je to hlavní, co bychom měli mít na paměti, když si vybíráme styly kódování a diskutujeme o nich.

> Čtení oblíbených stylových průvodců nám umožní držet krok s nejnovějšími idejemi o trendech kódovacích stylů a nejlepších praktikách.

Reviewer: ```suggestion
Čtení oblíbených stylových průvodců nám umožní držet krok s nejnovějšími myšlenami o trendech stylů zápisu programu a nejlepších praktikách.
```
